src/streaming/downsampler.cpp

```diff
@@ -0,0 +1,295 @@
+#include "downsampler.hh"
+#include "macros.hh"
+#include "zarrv2.array.writer.hh"
+#include "zarrv3.array.writer.hh"
+
+namespace {
+template<typename T>
+[[nodiscard]] ByteVector
+scale_image(ConstByteSpan src, size_t& width, size_t& height)
+{
+    const auto bytes_of_src = src.size();
+    const auto bytes_of_frame = width * height * sizeof(T);
+
+    EXPECT(bytes_of_src >= bytes_of_frame,
+           "Expecting at least ",
+           bytes_of_frame,
+           " bytes, got ",
+           bytes_of_src);
+
+    const int downscale = 2;
+    constexpr auto bytes_of_type = static_cast<double>(sizeof(T));
+    const double factor = 0.25;
+
+    const auto w_pad = static_cast<double>(width + (width % downscale));
+    const auto h_pad = static_cast<double>(height + (height % downscale));
+
+    const auto size_downscaled =
+      static_cast<uint32_t>(w_pad * h_pad * factor * bytes_of_type);
+
+    ByteVector dst(size_downscaled, static_cast<std::byte>(0));
+    auto* dst_as_T = reinterpret_cast<T*>(dst.data());
+    auto* src_as_T = reinterpret_cast<const T*>(src.data());
+
+    size_t dst_idx = 0;
+    for (auto row = 0; row < height; row += downscale) {
+        const bool pad_height = (row == height - 1 && height != h_pad);
+
+        for (auto col = 0; col < width; col += downscale) {
+            size_t src_idx = row * width + col;
+            const bool pad_width = (col == width - 1 && width != w_pad);
+
+            auto here = static_cast<double>(src_as_T[src_idx]);
+            auto right = static_cast<double>(
+              src_as_T[src_idx + (1 - static_cast<int>(pad_width))]);
+            auto down = static_cast<double>(
+              src_as_T[src_idx + width * (1 - static_cast<int>(pad_height))]);
+            auto diag = static_cast<double>(
+              src_as_T[src_idx + width * (1 - static_cast<int>(pad_height)) +
+                       (1 - static_cast<int>(pad_width))]);
+
+            dst_as_T[dst_idx++] =
+              static_cast<T>(factor * (here + right + down + diag));
```

Comment:
this is one of those cases where division might work better, since it avoids a cast to float for fixed point types.

`T(here+right+down+diag)/T(4)`

Reply:
Now that I think about this more, I don't know if this is even the correct approach. Consider a 2x2 grid with values 1, 1, 1, 0. The way I do it and the way you suggest would both truncated the averaged value to 0, which means we'd lose the dominant signal. Should we be rounding instead? We could do

```
dst_as_T[dst_idx++] = static_cast<T>((here + right + down + diag + 2) / 4);
```

which would effectively round to the nearest integer.